src/vos/vos_obj.c

```diff
@@ -668,7 +668,7 @@ vos_obj_del_key(daos_handle_t coh, daos_unit_oid_t oid, daos_key_t *dkey,
 	int			 rc;
 
 	rc = vos_obj_hold(occ, cont, oid, &epr, 0, VOS_OBJ_VISIBLE,
-			  DAOS_INTENT_KILL, &obj, NULL);
+			  DAOS_INTENT_PUNCH, &obj, NULL);
```

Comment:
What is the iterator doing here?   I presume KILL was used for a reason, such as the fact that it checks the reference count on the object before removing stuff from the tree.   Depending on the operations the iterator is performing, this type of thing could be unsafe.

As an example that probably doesn't apply to sysdb.  When aggregation comes back from a yield over an evtree, it doesn't re-probe because it knows the evtree isn't going anywhere and it has a cached state.   But an API like vos_obj_del_key would simply remove the whole evtree causing problems for the aggregation iterator.

The reference count and object eviction for KILL is overkill because what you really want is a reference count on the key tree but it would prevent scenarios such as these.   That said, it's probably not an issue for smd but at the very least should be documented in the API documentation.

Reply:
In DAOS check, we have table "chk" in sysdb, each pool has an akey under such table. Sometimes, we need to travel the table to delete the pools under such table.

Reply:
Yes, the chk logic iterates the table, and conditionally delete some keys from the table via iteration callback.